### ISSUE TITLE
Preview doesn't start, get error messages instead

### ISSUE THREAD

**Issue by deHarro** [JohnAtl/pcbgcode#15]
Hi John,

if I check "Show preview" in the setup ULP and then run "make my board", I don't get a preview but error messages as below:

![grafik](https://user-images.githubusercontent.com/26298406/185358949-4b09860f-c4f8-4cb4-8c89-c37934c876e0.png)

[edit]
Message translated: "There is no app associated with the file to perform this action. Install an app, or create an association on the Default Apps settings page if an app is already installed."
[\edit]

![grafik](https://user-images.githubusercontent.com/26298406/185359003-0a6c9ff0-c86c-4884-8ffc-54d6f5679f47.png)

This problem is not a new one (not correlated with the Alpha version of pcb-gcode), but sorrily I'm unable to give the exact time spot, when the problems first occured.

I already updated and lastly reinstalled Java and Processing, but to no avail.
I'm working with Windows 10, fully patched.

[edit2]
If I go to "...\pcb-gcode-3.7.0\viewer" and start "viewer.pde" and then select "run" manually, the preview is displayed correctly.
If I go to "...\pcb-gcode-3.7.0\viewer\applet" and start "viewer.jar" manually, the preview is displayed correctly.
[\edit2]

[edit3]
If I start "...\pcb-gcode-3.7.0\viewer\application.windows64\viewer.exe" I get displayed a demo layout.
The same with "...\pcb-gcode-3.7.0\viewer\application.windows32\viewer.exe".
On the other hand, if I start "...\pcb-gcode-3.7.0\viewer\application.windows\viewer.exe" the error message is displayed, as above.

The difference is apparently when looking at the icons of the different viewer apps:

![grafik](https://user-images.githubusercontent.com/26298406/185369675-ee642272-7e9f-4c06-87a1-fdecbc6ac239.png)
![grafik](https://user-images.githubusercontent.com/26298406/185369790-f506bab9-1666-4dab-864b-d946a2bd36f6.png)
![grafik](https://user-images.githubusercontent.com/26298406/185369934-12c56a0b-0c7a-4d76-97d7-49b872fe3503.png)

Obviously there is no file association with the viewer.exe in the "windows" folder, and apparently, the ULP calls this not associated viewer.exe.

If I edit the calling line 132 in "pcb-gcode.ulp" from       
`system(g_path + "/viewer/application.windows/viewer.exe");`
to read
`system(g_path + "/viewer/application.windows32/viewer.exe");`
or
`system(g_path + "/viewer/application.windows64/viewer.exe");`
the error message is not displayed, but there is also no preview window :(

So, where can I define which viewer to use?
[\edit3]

Thanks!

Harald

**Comment by deHarro** [comment]
Additional info:
If I load "viewer.pde" from the windows folder (the one I referred to in my last post above, "\windows", not "\windows32" or "\windows64") into Proccessing IDE, I get a warning about replacement of the "processing.xml" library by an inbuild XML class -> "processing.xml no longer exists, this code needs to be updated for 2.0."

Any help appreciated...
Harald

**Comment by deHarro** [comment]
Hi John!

I succeeded in getting the preview running again.

The problem was as follows:
There is some sort of "shunting yard" for java where symbolic links (symlinks) point to the location where the actual java executables reside.
Presumably this shunting yard (actually "C:\ProgramData\Oracle\Java\javapath\...") was a remainder of an old java installation and pointed to a nonexistent folder location.
Since symlinks cannot be edited, I had to remove and then recreate the symlink to javaw.exe in this shunting yard location.

The recreation was done with
```
C:\WINDOWS\system32>mklink C:\ProgramData\Oracle\Java\javapath\javaw.exe "C:\Program Files (x86)\Java\jre1.8.0_341\bin\javaw.exe"
symbolische Verknüpfung erstellt für C:\ProgramData\Oracle\Java\javapath\javaw.exe <<===>> C:\Program Files (x86)\Java\jre1.8.0_341\bin\javaw.exe
```
Now preview works as expected again.

... hopefully subsequent updates of Java can work with this newly generated symlink :)

I close the issue.

Harald
